hw/ip/otp_ctrl/doc/_index.md

```diff
@@ -25,30 +296,418 @@ The following table lists the main parameters used throughout the OTP controller
+
+Note that the register space exposed via the TL-UL interface is dependent on the underlying proprietary OTP IP and hence not further described in this document.
+
+#### Generalized Open-source Interface
+
+The generalized open-source interface uses a couple of parameters (defaults set for Earlgrey configuration).
+
+Parameter      | Default | Top Earlgrey | Description
+---------------|---------|--------------|---------------
+`Width`        | 16      | 16           | Payload data width.
+`Depth`        | 1024    | 1024         | Depth of OTP macro.
+`CmdWidth`     | 2       | 2            | Width of the OTP command.
+`ErrWidth`     | 4       | 4            | Width of error code output signal.
+
+The generalized open-source interface is a simple command interface with a ready / valid handshake that makes it possible to introduce back pressure if the OTP macro is not able to accept a command due to an ongoing operation.
+
+Signal                  | Direction        | Type                        | Description
+------------------------|------------------|-----------------------------|---------------
+`ready_o`               | `output`         | `logic`                     | Ready signal for the command handshake.
+`valid_i`               | `input`          | `logic`                     | Valid signal for the command handshake.
+`cmd_i`                 | `input`          | `logic [CmdWidth-1:0]`      | OTP command: `2'b00 = read`, `2'b01 = write`, `2'b11 = initialize`
+`addr_i`                | `input`          | `logic [$clog2(Depth)-1:0]` | OTP word address.
+`wdata_i`               | `input`          | `logic [Width-1:0]`         | Write data for write commands.
+`valid_o`               | `output`         | `logic`                     | Valid signal for command response.
+`rdata_o`               | `output`         | `logic [Width-1:0]`         | Read data from read commands.
+`err_o`                 | `output`         | `logic [ErrWidth-1:0]`      | Error code.
+
+Responses are returned in-order via an unidirectional response interface (i.e., without back pressure capability).
+Downstream logic must be able to sink the response in any case. The response optionally carries read data, depending on whether the operation that took place was a read or not.
+Also, an error signal returns a non-zero error code in case an error occurred while carrying out the OTP command.
+The error codes are defined further below.
+
+Error Code | Description
+-----------|-------------
+0x0        | Command completed successfully.
+0x1        | Invalid command.
+0x2        | Read or write attempted before initialization.
+0x3        | Failure during OTP initialization.
+0x4        | Correctable read error occurred.
+0x5        | Uncorrectable read error occurred.
+0x6        | Other read errors (TBD).
+0x7        | Programming error occurred (TBD).
+others     | Reserved.
```

Comment:
This err_code seems to be different from the ERR_CODE section in OTP register.

Reply:
yeah they do not map one-to-one at the moment.

we have to finetune these error codes once the inner workings of the closed source wrapper are clearer. at the moment, these codes just serve as a placeholder. I'll add a TODO here mentioning this.